control-plane-operator/controllers/hostedcontrolplane/cloud/azure/azure.go

```diff
@@ -8,10 +8,44 @@ import (
 	corev1 "k8s.io/api/core/v1"
 )
 
-const CloudConfigKey = "cloud.conf"
+const (
+	CloudConfigKey = "cloud.conf"
+	Provider       = "azure"
+)
 
 func ReconcileCloudConfig(cm *corev1.ConfigMap, hcp *hyperv1.HostedControlPlane, credentialsSecret *corev1.Secret) error {
-	cfg := AzureConfig{
+	cfg := azureConfigWithoutCredentials(hcp, credentialsSecret)
+	serializedConfig, err := json.MarshalIndent(cfg, "", "  ")
+	if err != nil {
+		return fmt.Errorf("failed to serialize cloudconfig: %w", err)
+	}
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+	cm.Data[CloudConfigKey] = string(serializedConfig)
+
+	return nil
+}
+
+func ReconcileCloudConfigWithCredentials(secret *corev1.Secret, hcp *hyperv1.HostedControlPlane, credentialsSecret *corev1.Secret) error {
+	cfg := azureConfigWithoutCredentials(hcp, credentialsSecret)
+	cfg.AADClientID = string(credentialsSecret.Data["AZURE_CLIENT_ID"])
+	cfg.AADClientSecret = string(credentialsSecret.Data["AZURE_CLIENT_SECRET"])
```

Comment:
how do we know secret has all valid/expected keys? are we validating this some where?

Reply:
same question for AZURE_TENANT_ID

Reply:
It'd be nice indeed to have some units for ReconcileCloudConfigWithCredentials covering this cases

Reply:
So checking if these keys exist makes sense but I would do that at a much higher level, e.G. in the hostedcluster controller, does that sound good to you?

What kind of regression are you hoping to find by testing `ReconcileCloudConfigWithCredentials` with an invalid secret? It will just result in a azure cloud config that has the empty string for `AADClientID` and `AADClientSecret`

Reply:
>So checking if these keys exist makes sense but I would do that at a much higher level, e.G. in the hostedcluster controller, does that sound good to you?

that sounds great to me.

>What kind of regression are you hoping to find by testing ReconcileCloudConfigWithCredentials with an invalid secret? It will just result in a azure cloud config that has the empty string for AADClientID and AADClientSecret

that's right, so I'd expect us to fail if the secret is not as expected, so if we do the validation as mentioned above and unit test that I think that's good enough.
My thinking is the same than in https://github.com/openshift/hypershift/pull/1055#discussion_r814223962, we should ensure we programatically prevent from happening any "unsupported" scenario

Reply:
like right now by not validating we let users shoot themselves in the foot by having a typo in their config, that would result in undefined behaviour and we wouldn't be signalling that meaningfully anywhere.